File: 0383-ransom-note/0383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
const canConstruct = (ransomNote, magazine) => {
    const map = {};
    for (let i = 0; i < magazine.length; i++) {
        map[magazine[i]] = map[magazine[i]] + 1 || 1;
    }
    
    for (let i = 0; i < ransomNote.length; i++) {
        if (!map[ransomNote[i]]) return false;
        else map[ransomNote[i]]--;
    }
    return true;
};

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        countRansom = {}
        countMagazine = {}
        for char in ransomNote:
            if char not in countRansom:
                countRansom[char] = 1
            else:
                countRansom[char] += 1
        for char in magazine:
            if char not in countMagazine:
                countMagazine[char] = 1
            else:
                countMagazine[char] += 1
        for key in countRansom:
            if key not in countMagazine or countRansom[key] > countMagazine[key]:
                return False
        return True